src/LinkedGenome.py

```diff
@@ -0,0 +1,192 @@
+    
+    def __iter__(self):
+        feature = self.head.next
+        while feature is not self.head:
+            yield feature.val
+            feature = feature.next
+    def into_iter_with_pos(self):
+        feature = self.head.next
+        acc = 0
+        while feature is not self.head:
+            acc += feature.val.length 
+            yield (feature, acc)
+            feature = feature.next
+
+    def insert_te(self, pos: int, length: int) -> int:
+        """
+        Insert a new transposable element.
+
+        Insert a new transposable element at position pos and len
+        nucleotide forward.
+
+        If the TE collides with an existing TE, i.e. genome[pos]
+        already contains TEs, then that TE should be disabled and
+        removed from the set of active TEs.
+
+        Returns a new ID for the transposable element.
+        """
+
+        for feature, end in self.into_iter_with_pos():
+            if end > pos:
+                self.insert_into(length, feature, feature.val.length -end + pos, end - pos)
+                return self.counter_te
+        return -1
+
+    def insert_into(self, length, feature, split_size_1, split_size_2):
+        if  feature.val.feature == self.active_te:
+            self.disable_feature(feature)
+        feature.val = Feature(feature.val.feature, split_size_1)
+        insert_after(feature, Feature(self.active_te, length))
+        insert_after(feature.next, Feature(feature.val.feature, split_size_2))
+        self.counter_te += 1
+        self.active_identifier[self.counter_te] = feature.next
+
+    def disable_feature(self, feature):
+        feature.val = Feature(self.inactive_te, feature.val.length)
+        for identifier, disable in self.active_identifier.items():
+            if disable is feature:
+                self.active_identifier.pop(identifier)
+                break
+
+    def copy_te(self, te: int, offset: int) -> int | None:
+        """
+        Copy a transposable element.
+
+        Copy the transposable element te to an offset from its current
+        location.
+
+        The offset can be positive or negative; if positive the te is copied
+        upwards and if negative it is copied downwards. If the offset moves
+        the copy left of index 0 or right of the largest index, it should
+        wrap around, since the genome is circular.
+
+        If te is not active, return None (and do not copy it).
+        """
+        original = feature= self.active_identifier[te]
+        if original.val.feature != self.active_te:
+            return None
+        match offset > 0:
+            case True:
+                direction = "next"
+                offset = offset - original.val.length
+            case False:
+                direction = "prev"
+                offset *= -1
+        while offset > 0:
+            feature = getattr(feature, direction)
+            if feature is self.head:
+                feature = getattr(feature, direction)
```

Comment:
The use of `getattr` to go in the correct direction is a bit cheeky, but I like it!